tiny_dnn/layers/global_average_pooling_layer.h

```diff
@@ -0,0 +1,130 @@
+/*
+    Copyright (c) 2015, Taiga Nomi
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "tiny_dnn/layers/layer.h"
+#include "tiny_dnn/util/util.h"
+
+#include "tiny_dnn/core/kernels/global_avepool_grad_op.h"
+#include "tiny_dnn/core/kernels/global_avepool_op.h"
+
+namespace tiny_dnn {
+
+/**
+ * applies channel-wise global average pooling to spatial data.
+ **/
+class global_average_pooling_layer : public layer {
+ public:
+  using layer::parallelize_;
+
+  global_average_pooling_layer(const shape3d &in_shape,
+                               backend_t backend_type = core::default_engine())
+    : global_average_pooling_layer(
+        in_shape.width_, in_shape.height_, in_shape.depth_, backend_type) {}
```

Comment:
Do we need separate `width` and `height`? We can generalize to any number of dimensions by passing just area and depth, without caring on number of dimensions. Or passing just depth and deducing area. Or even deducing all the parameters similarly to activations.